pythoncode/usbTrainer.py

```diff
@@ -92,477 +101,826 @@
-                if dev != None:
-                    msg = "GetTrainer - Fortius head unit initialised"
-                    trainer_type = tt_FortiusSB
-                else:
-                    msg = "GetTrainer - Unable to load firmware"
-                    dev = False
-            except:                                                     # not found
-                msg = "GetTrainer - Unable to initialise trainer"
-                dev = False
+    def SendToTrainer(self, TacxMode):
+        assert (TacxMode in (modeStop, modeResistance, modeCalibrate))
 
-        #-----------------------------------------------------------------------
-        # Set configuration
-        #-----------------------------------------------------------------------
-        if dev != False:
-            dev.set_configuration()
-            if trainer_type == tt_iMagic:   dev.set_interface_altsetting(0, 1)
+        Calibrate = self.Calibrate
+        PedalEcho = self.PedalEcho
+        Target    = self.TargetResistance
+        Weight    = 0
+
+        if debug.on(debug.Function):
+            logfile.Write ("clsTacxTrainer.SendToTrainer(T=%s, M=%s, P=%s, G=%s, R=%s, W=%s, PE=%s, S=%s, C=%s)" % \
+            (TacxMode, self.TargetMode, self.TargetPower, self.TargetGrade, \
+            self.TargetResistance, self.UserAndBikeWeight, self.PedalEcho, \
+            self.SpeedKmh, self.Cadence))
 
         #-----------------------------------------------------------------------
-        # InitialiseTrainer (will not read cadence until initialisation byte is sent)
+        # Prepare parameters to be sent to trainer
         #-----------------------------------------------------------------------
-        if dev != False and LegacyProtocol == False:
-            data = struct.pack (sc.unsigned_int, 0x00000002)
-            if debug.on(debug.Data2):
-                logfile.Write ("InitialiseTrainer data=%s (len=%s)" % (logfile.HexSpace(data), len(data)))
-            dev.write(0x02,data)
+        error = False
+        if  TacxMode == modeStop:
+            Calibrate   = 0
+            PedalEcho   = 0
+            Target      = 0
+            Weight      = 0
+
+        elif TacxMode == modeResistance:
+            if Calibrate == 0:                                  # Use old formula:
+                Calibrate   = 0                                 # may be -8...+8
+                Calibrate   = ( Calibrate + 8 ) * 130           # 0x0410
+
+            if self.TargetMode == mode_Power:
+                Target = self.TargetResistance
+                Weight = 0x0a                                   # weight=0x0a is a good fly-wheel value
+                                                                # UserAndBikeWeight is not used!
+            elif self.TargetMode == mode_Grade:
+                Target = self.TargetResistance
+                Weight = 0x0a                                   # weight=0x0a is a good fly-wheel value

```

Comment:
Disagree 0x0a is a good value

Reply:
Suggestion: don't what is should not be, without stating what it should :-)

Reply:
Flywheel weight should be UserAndBikeWeight for simulation trainer mode (ie mode_Grade). I really want to know why "an 100kg flywheels gives undesired behaviour" because for me, it gives much (much) better behaviour and a 10kg flywheel gives undesired behaviour. Test is.... ride downhill and stop pedalling, do you carry your speed? Ride on flat and hit a sudden 4% downhill, does resistance drop immediately, or do you gather speed? Ride on flat and hit a sudden 4% uphill (Rouvy is great at sudden changes!), does resistance hit a wall or do you slowly lose momentum.

Reply:
Cool. Will implement and test.
Interesting info for @totalreverse for the interface description.
Thanks

Reply:
I was riding my trainer on Rouvy at the time I made the change, restarted FortiusAnt and wrote that comment, while still pedalling. ;)

Reply:
I see that @totalreverse changed field-description jan 4th. Makes sense.
antifier has been struggling with this as well.

Reply:
> Ride on flat and hit a sudden 4% uphill (Rouvy is great at sudden changes!),
Fully agree.

Reply:
Zwift can send fractional kg, so ensure int() when populating the message to the trainer.

Reply:
done